src/fu-udev-device.c

```diff
@@ -108,29 +108,40 @@ fu_udev_device_get_sysfs_attr_as_uint8 (GUdevDevice *udev_device, const gchar *n
 }
 
 static void
-fu_udev_device_dump_internal (GUdevDevice *udev_device)
+fu_udev_device_verbose (FuDevice *device, guint idt, GString *str)
 {
+	FuUdevDevice *self = FU_UDEV_DEVICE (device);
+	FuUdevDevicePrivate *priv = GET_PRIVATE (self);
 	const gchar * const *keys;
 
-	keys = g_udev_device_get_property_keys (udev_device);
+	if (priv->udev_device == NULL)
+		return;
+
+	keys = g_udev_device_get_property_keys (priv->udev_device);
 	for (guint i = 0; keys[i] != NULL; i++) {
-		g_debug ("%s={%s}", keys[i],
-			 g_udev_device_get_property (udev_device, keys[i]));
+		fu_common_string_append_kv (str, idt, keys[i],
+					    g_udev_device_get_property (priv->udev_device,
+									keys[i]));
 	}
-	keys = g_udev_device_get_sysfs_attr_keys (udev_device);
+	keys = g_udev_device_get_sysfs_attr_keys (priv->udev_device);
 	for (guint i = 0; keys[i] != NULL; i++) {
-		g_debug ("%s=[%s]", keys[i],
-			 g_udev_device_get_sysfs_attr (udev_device, keys[i]));
+		fu_common_string_append_kv (str, idt, keys[i],
+					    g_udev_device_get_sysfs_attr (priv->udev_device,
+									  keys[i]));
 	}
 }
 
-void
-fu_udev_device_dump (FuUdevDevice *self)
+static void
+fu_udev_device_to_string (FuDevice *device, guint idt, GString *str)
 {
-	FuUdevDevicePrivate *priv = GET_PRIVATE (self);
-	if (priv->udev_device == NULL)
-		return;
-	fu_udev_device_dump_internal (priv->udev_device);
+	FuDeviceClass *klass = FU_DEVICE_GET_CLASS (device);
+
+	if (g_getenv ("FU_UDEV_DEVICE_DEBUG") != NULL)
+		fu_udev_device_verbose (device, idt, str);
+
+	/* subclassed */
+	if (klass->to_string != NULL)
+		klass->to_string (device, idt + 1, str);
```

Comment:
That's going to be an infinite loop. Just drop those three lines as the subclass won't need the default version.

Reply:
Makes sense now that I think about it.  I was thinking that devices subclassing would be able to have their own to_string's but they just override this one.

Fixed.